File: app/src/main/java/com/example/nasmovie/util/FileUtils.java
package com.example.nasmovie.util;

import java.io.File;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.DecimalFormat;

/**
 * 文件工具类
 */
public class FileUtils {

    /**
     * 获取文件扩展名
     */
    public static String getExtension(String filename) {
        if (filename == null || !filename.contains(".")) {
            return "";
        }
        return filename.substring(filename.lastIndexOf(".") + 1).toLowerCase();
    }

    /**
     * 获取不带扩展名的文件名
     */
    public static String getNameWithoutExtension(String filename) {
        if (filename == null) {
            return "";
        }
        int dotIndex = filename.lastIndexOf(".");
        return dotIndex > 0 ? filename.substring(0, dotIndex) : filename;
    }

    /**
     * 格式化文件大小
     */
    public static String formatFileSize(long bytes) {
        if (bytes < 1024) {
            return bytes + " B";
        } else if (bytes < 1024 * 1024) {
            return String.format("%.1f KB", bytes / 1024.0);
        } else if (bytes < 1024 * 1024 * 1024) {
            return String.format("%.1f MB", bytes / (1024.0 * 1024));
        } else {
            DecimalFormat df = new DecimalFormat("#.##");
            return df.format(bytes / (1024.0 * 1024 * 1024)) + " GB";
        }
    }

    /**
     * 格式化时长（毫秒转为时分秒）
     */
    public static String formatDuration(long milliseconds) {
        int seconds = (int) (milliseconds / 1000);
        int hours = seconds / 3600;
        int minutes = (seconds % 3600) / 60;
        int secs = seconds % 60;

        if (hours > 0) {
            return String.format("%d:%02d:%02d", hours, minutes, secs);
        } else {
            return String.format("%02d:%02d", minutes, secs);
        }
    }

    /**
     * 格式化时长（分钟转为时分）
     */
    public static String formatDurationMinutes(int minutes) {
        int hours = minutes / 60;
        int mins = minutes % 60;

        if (hours > 0) {
            return String.format("%d小时%d分钟", hours, mins);
        } else {
            return String.format("%d分钟", mins);
        }
    }

    /**
     * 计算字符串的MD5哈希值
     */
    public static String md5(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] digest = md.digest(input.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) {
                sb.append(String.format("%02x", b & 0xff));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            return String.valueOf(input.hashCode());
        }
    }

    /**
     * 生成电影ID（基于路径）
     */
    public static String generateMovieId(String path) {
        return md5(path);
    }

    /**
     * 判断是否为视频文件
     */
    public static boolean isVideoFile(String filename) {
        String ext = getExtension(filename);
        return "mp4".equals(ext) || "mkv".equals(ext);
    }

    /**
     * 判断是否为字幕文件
     */
    public static boolean isSubtitleFile(String filename) {
        String ext = getExtension(filename);
        return "srt".equals(ext) || "ass".equals(ext) || "ssa".equals(ext)
            || "sub".equals(ext) || "vtt".equals(ext);
    }

    /**
     * 判断是否为海报图片
     */
    public static boolean isPosterImage(String filename) {
        if (filename == null) return false;
        String lowerName = filename.toLowerCase();
        return "poster.jpg".equals(lowerName) || "poster.png".equals(lowerName)
            || "folder.jpg".equals(lowerName) || "folder.png".equals(lowerName)
            || lowerName.startsWith("poster.") || lowerName.startsWith("folder.");
    }

    /**
     * 判断是否为NFO文件
     */
    public static boolean isNfoFile(String filename) {
        return "nfo".equals(getExtension(filename));
    }

    /**
     * 组合SMB路径
     */
    public static String combineSmbPath(String... parts) {
        StringBuilder sb = new StringBuilder();
        for (String part : parts) {
            if (part == null || part.isEmpty()) continue;

            if (sb.length() > 0 && !sb.toString().endsWith("/") && !sb.toString().endsWith("\\")) {
                sb.append("/");
            }

            // 移除开头的斜杠
            if (part.startsWith("/") || part.startsWith("\\")) {
                part = part.substring(1);
            }

            sb.append(part);
        }
        return sb.toString();
    }
}

File: app/src/main/java/com/example/nasmovie/util/FileUtils.kt
package com.example.nasmovie.util

import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.text.DecimalFormat
import java.util.Locale

/**
 * 文件工具类
 */
object FileUtils {

    /**
     * 获取文件扩展名
     */
    fun getExtension(filename: String?): String {
        if (filename == null || !filename.contains(".")) {
            return ""
        }
        return filename.substring(filename.lastIndexOf(".") + 1).lowercase(Locale.ROOT)
    }

    /**
     * 获取不带扩展名的文件名
     */
    fun getNameWithoutExtension(filename: String?): String {
        if (filename == null) {
            return ""
        }
        val dotIndex = filename.lastIndexOf(".")
        return if (dotIndex > 0) filename.substring(0, dotIndex) else filename
    }

    /**
     * 格式化文件大小
     */
    fun formatFileSize(bytes: Long): String {
        return when {
            bytes < 1024 -> "$bytes B"
            bytes < 1024 * 1024 -> String.format(Locale.getDefault(), "%.1f KB", bytes / 1024.0)
            bytes < 1024 * 1024 * 1024 -> String.format(Locale.getDefault(), "%.1f MB", bytes / (1024.0 * 1024))
            else -> {
                val df = DecimalFormat("#.##")
                df.format(bytes / (1024.0 * 1024 * 1024)) + " GB"
            }
        }
    }

    /**
     * 格式化时长（毫秒转为时分秒）
     */
    fun formatDuration(milliseconds: Long): String {
        val seconds = (milliseconds / 1000).toInt()
        val hours = seconds / 3600
        val minutes = seconds % 3600 / 60
        val secs = seconds % 60

        return if (hours > 0) {
            String.format(Locale.getDefault(), "%d:%02d:%02d", hours, minutes, secs)
        } else {
            String.format(Locale.getDefault(), "%02d:%02d", minutes, secs)
        }
    }

    /**
     * 格式化时长（分钟转为时分）
     */
    fun formatDurationMinutes(minutes: Int): String {
        val hours = minutes / 60
        val mins = minutes % 60

        return if (hours > 0) {
            String.format(Locale.getDefault(), "%d小时%d分钟", hours, mins)
        } else {
            String.format(Locale.getDefault(), "%d分钟", mins)
        }
    }

    /**
     * 计算字符串的MD5哈希值
     */
    fun md5(input: String): String {
        return try {
            val md = MessageDigest.getInstance("MD5")
            val digest = md.digest(input.toByteArray())
            digest.joinToString("") { "%02x".format(it) }
        } catch (e: NoSuchAlgorithmException) {
            input.hashCode().toString()
        }
    }

    /**
     * 生成电影ID（基于路径）
     */
    fun generateMovieId(path: String): String = md5(path)

    /**
     * 判断是否为视频文件
     */
    fun isVideoFile(filename: String?): Boolean {
        val ext = getExtension(filename)
        return ext in AppConstants.VIDEO_EXTENSIONS
    }

    /**
     * 判断是否为字幕文件
     */
    fun isSubtitleFile(filename: String?): Boolean {
        val ext = getExtension(filename)
        return ext in AppConstants.SUBTITLE_EXTENSIONS
    }

    /**
     * 判断是否为海报图片
     */
    fun isPosterImage(filename: String?): Boolean {
        if (filename == null) return false
        val lowerName = filename.lowercase(Locale.ROOT)
        return lowerName in listOf(
            "poster.jpg", "poster.png",
            "folder.jpg", "folder.png"
        ) || lowerName.startsWith("poster.") || lowerName.startsWith("folder.")
    }

    /**
     * 判断是否为NFO文件
     */
    fun isNfoFile(filename: String?): Boolean = getExtension(filename) == "nfo"

    /**
     * 组合SMB路径
     */
    fun combineSmbPath(vararg parts: String?): String {
        val sb = StringBuilder()
        for (part in parts) {
            if (part.isNullOrEmpty()) continue

            if (sb.isNotEmpty() && !sb.endsWith("/") && !sb.endsWith("\\")) {
                sb.append("/")
            }

            // 移除开头的斜杠
            val trimmedPart = part.removePrefix("/").removePrefix("\\")

            sb.append(trimmedPart)
        }
        return sb.toString()
    }
}